src/inmanta/data/__init__.py

```diff
@@ -4031,6 +4093,153 @@ def to_dict(self) -> JsonType:
+    """ Represents the ordering by which resources should be sorted"""
+
+    validator_dataclass = Resource
+    valid_sort_columns = {"resource_type": str, "agent": str, "status": str, "resource_id_value": str}
+
+    @classmethod
+    def parse_from_string(
+        cls,
+        sort: str,
+    ) -> "ResourceOrder":
+        database_order = super().parse_from_string(sort)
+        return ResourceOrder(database_order.order_by_column, database_order.order)
+
+    def get_order_by_db_column_name(self) -> ColumnNameStr:
+        return ColumnNameStr(
+            f"{super().get_order_by_db_column_name()}{'::text' if self._should_be_treated_as_string() else ''}"
+        )
+
+    def _should_be_treated_as_string(self) -> bool:
+        """Ensure that records are sorted alphabetically by status instead of the enum order"""
+        return self.order_by_column == "status"
+
+
+class ResourceHistoryOrder(DatabaseOrder):
+    """ Represents the ordering by which resource history should be sorted"""
+
+    # Sorting based on the date of the configuration model
+    validator_dataclass = ConfigurationModel
+    valid_sort_columns = {"date": datetime.datetime}
+
+    @classmethod
+    def parse_from_string(
+        cls,
+        sort: str,
+    ) -> "ResourceHistoryOrder":
+        database_order = super().parse_from_string(sort)
+        return ResourceHistoryOrder(database_order.order_by_column, database_order.order)
+
+
+class ResourceLogOrder(DatabaseOrder):
+    """ Represents the ordering by which resource logs should be sorted"""
+
+    validator_dataclass = ResourceAction
+    valid_sort_columns = {"timestamp": datetime.datetime}
+
+    @classmethod
+    def parse_from_string(
+        cls,
+        sort: str,
+    ) -> "ResourceLogOrder":
+        database_order = super().parse_from_string(sort)
+        return ResourceLogOrder(database_order.order_by_column, database_order.order)
+
+
+class CompileReportOrder(DatabaseOrder):
+    """ Represents the ordering by which compile reports should be sorted"""
+
+    validator_dataclass = Compile
+    valid_sort_columns = {"requested": datetime.datetime}
+
+    @classmethod
+    def parse_from_string(
+        cls,
+        sort: str,
+    ) -> "CompileReportOrder":
+        database_order = super().parse_from_string(sort)
+        return CompileReportOrder(database_order.order_by_column, database_order.order)
+
+
+class QueryBuilder:
+    def __init__(self, base_query: str, *values: List[object], initial_filter_statements: Optional[List[str]] = None) -> None:
+        self.sql_query = base_query
+        self.values = list(values)
+        if initial_filter_statements:
+            self.sql_query += BaseDocument._join_filter_statements(initial_filter_statements)
+        self.filter_started = bool(initial_filter_statements)
+
+    @property
+    def offset(self):
+        return len(self.values) + 1
+
+    def filter(self, filter_statements: List[str], values: List[object]) -> "QueryBuilder":
```

Comment:
I would make this class immutable and create a new QueryBuilder on each modification, to make them composable

Reply:
This is still very relevant